src/ott/neural/flows/otfm.py

```diff
@@ -0,0 +1,371 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import functools
+import types
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Literal,
+    Mapping,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+)
+
+import jax
+import jax.numpy as jnp
+
+import diffrax
+import optax
+from flax.training import train_state
+from orbax import checkpoint
+
+from ott import utils
+from ott.geometry import costs
+from ott.neural.flows.flows import BaseFlow
+from ott.neural.models.base_solver import (
+    BaseNeuralSolver,
+    ResampleMixin,
+    UnbalancednessMixin,
+)
+from ott.solvers import was_solver
+
+__all__ = ["OTFlowMatching"]
+
+
+class OTFlowMatching(UnbalancednessMixin, ResampleMixin, BaseNeuralSolver):
+  """(Optimal transport) flow matching class.
+
+  Flow matching as introduced in :cite:`lipman:22`, with extension to OT-FM
+  (:cite`tong:23`, :cite:`pooladian:23`).
+
+  Args:
+    velocity_field: Neural vector field parameterized by a neural network.
+    input_dim: Dimension of the input data.
+    cond_dim: Dimension of the conditioning variable.
+    iterations: Number of iterations.
+    valid_freq: Frequency of validation.
+    ot_solver: OT solver to match samples from the source and the target
+      distribution as proposed in :cite:`tong:23`, :cite:`pooladian:23`.
+      If :obj:`None`, no matching will be performed as proposed in
+      :cite:`lipman:22`.
+    flow: Flow between source and target distribution.
+    time_sampler: Sampler for the time.
+    optimizer: Optimizer for `velocity_field`.
+    checkpoint_manager: Checkpoint manager.
+    epsilon: Entropy regularization term of the OT OT problem solved by the
+      `ot_solver`.
+    cost_fn: Cost function for the OT problem solved by the `ot_solver`.
+    scale_cost: How to scale the cost matrix for the OT problem solved by the
+      `ot_solver`.
+    tau_a: If :math:`<1`, defines how much unbalanced the problem is
+      on the first marginal.
+    tau_b: If :math:`< 1`, defines how much unbalanced the problem is
+      on the second marginal.
+    rescaling_a: Neural network to learn the left rescaling function as
+      suggested in :cite:`eyring:23`. If :obj:`None`, the left rescaling factor
+      is not learnt.
+    rescaling_b: Neural network to learn the right rescaling function as
+      suggested in :cite:`eyring:23`. If :obj:`None`, the right rescaling factor
+      is not learnt.
+    unbalanced_kwargs: Keyword arguments for the unbalancedness solver.
+    callback_fn: Callback function.
+    num_eval_samples: Number of samples to evaluate on during evaluation.
+    rng: Random number generator.
+  """
+
+  def __init__(
+      self,
+      velocity_field: Callable[[
```

Comment:
I think instantiating this class would be clearer like this (the module names are ad-hoc):
```python
tau_a = ...
tau_b = ...

vf = models.VelocityField(input_dim=..., cond_dim=..., optimizer=...)
flow = flows.ConstantFlow(...)
time_sampler = layers.UniformSampler(...)

matching_fn = layers.OTMatcher(epsilon=..., scale_cost=..., tau_a=tau_a, tau_b=tau_b)  # or None

source_rescaler = layers.Rescaler(tau=tau_a, ...)  # or None
target_rescaler = layers.Rescaler(tau=tau_b, ...)  # or None

fm = otfm.OTFM(vf, flow=flow, time_sampler=time_sampler,
               source_rescaler=source_rescaler, target_rescaler=target_rescaler)